### DISCUSSION TITLE
Vertico distorts email attachments

### DISCUSSION THREAD

**Discussion by nameiwillforget** [minad/vertico#454] [Q&A] [answered]
When I use C-c C-a in Gnus with Vertico to attach, say, a plaintext file, I receive the following output:
```
<#part type=#("text/plain" 0 10 (fontified nil)) filename="~/gdb.txt" disposition=inline description=gdb>
<#/part>
```
This triggers this error:
```
smtpmail-send-it: Sending failed: 554 5.0.0 Error: failed to parse message: failed to get content type for for child 0:mime: expected slash after first token in response to .
```
When I turn off Vertico, it removes the brackets around the type description and sending the email works.

**Comment by minad** [comment]
Alexander Prähauser ***@***.***> writes:

> When I use C-c C-a in Gnus with Vertico to attach, say, a plaintext file, I receive the following output:
> ```
> <#part type=#("text/plain" 0 10 (fontified nil)) filename="~/gdb.txt" disposition=inline description=gdb>
> <#/part>
> ```
> This triggers this error:
> ```
> smtpmail-send-it: Sending failed: 554 5.0.0 Error: failed to parse message: failed to get content type for for child 0:mime: expected slash after first token in response to .
> ```
> When I turn off Vertico, it removes the brackets around the type description and sending the email works.
>>
Let me try:


Test file

**Comment by minad** [comment]
In my test on Emacs 29.1, I could not reproduce the problem. Can you reproduce the issue with emacs -Q and a minimal configuration with only Vertico installed? Which exact key presses do you invoke?

**Reply by minad** [reply, marked as answer]
Maybe you've set `minibuffer-allow-text-properties=t` in your configuration? Compare:

~~~ elisp
;; OK
(let ((minibuffer-allow-text-properties nil))
  (completing-read "Select: " 
    (mapcar (lambda (x) (propertize x 'face 'error)) '("alpha" "beta" "gamma")))) ;; Returns string without properties

;; BUG
(let ((minibuffer-allow-text-properties t))
  (completing-read "Select: " 
    (mapcar (lambda (x) (propertize x 'face 'error)) '("alpha" "beta" "gamma")))) ;; Returns propertized string
~~~

**Comment by nameiwillforget** [comment]
"Daniel Mendler" ***@***.***> writes:

> Maybe you've set minibuffer-allow-text-properties=t in your configuration? Compare:
>
> (let ((minibuffer-allow-text-properties nil))
>   (completing-read "Select: " (mapcar (lambda (x) (propertize x 'face 'error)) '("alpha" "beta" "gamma")))) ;; Returns string without properties
>
> (let ((minibuffer-allow-text-properties t))
>   (completing-read "Select: " (mapcar (lambda (x) (propertize x 'face 'error)) '("alpha" "beta" "gamma")))) ;; Returns propertized string
>
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you authored the thread.

Ah, that was set to t, and apparantly by isearch+. Still, should there
maybe be a guard against that or something?

**Reply by minad** [reply]
> Ah, that was set to t, and apparantly by isearch+. Still, should there
maybe be a guard against that or something?

The guard is in `read-from-minibuffer`, which strips the text properties and which is used internally by `completing-read`. If you set `minibuffer-allow-text-properties` globally to t, this function is defeated.

This is not a bug. The variable should not be modified globally. Note that you can also reproduce this problem without Vertico in default completion. Run the following code:

~~~ elisp
(let ((minibuffer-allow-text-properties t))
  (completing-read "Select: " 
    (mapcar (lambda (x) (propertize x 'face 'error)) '("alpha" "beta" "gamma"))))
~~~

Then press the page down key which invokes `next-history-element`. This will then insert the colorized `#("alpha" 0 5 (face error))`. Press RET and `completing-read` will return a propertized string.

**Reply by minad** [reply]
While I insist that it is a mistake to set `minibuffer-allow-text-properties` globally, I still wonder why Gnus uses propertized candidates in the first place, in particular with a useless `fontified=nil` property. Ideally commands which pass propertized candidates can also handle propertized return values. So maybe a guard could be added to Gnus, but I would rather recommend to strip the properties properly before passing the candidates like text/plain to `gnus-completing-read`.

**Comment by nameiwillforget** [comment]
"Daniel Mendler" ***@***.***> writes:

>  Ah, that was set to t, and apparantly by isearch+. Still, should there
>  maybe be a guard against that or something?
>
> The guard is in read-from-minibuffer, which strips the text properties and which is used internally by completing-read. If you
> set minibuffer-allow-text-properties globally to t, this function is defeated.
>
> This is not a bug. The variable should not be modified globally. Note that you can also reproduce this problem without Vertico
> in default completion. Run the following code:
>
> (let ((minibuffer-allow-text-properties t))
>   (completing-read "Select: "
>     (mapcar (lambda (x) (propertize x 'face 'error)) '("alpha" "beta" "gamma"))))
>
> Then press the page down key which invokes next-history-element. This will then insert the colorized #("alpha" 0 5 (face
> error)). Then press RET and completing-read will return with a propertized string.
>
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you authored the thread.

Huh, but it doesn't. When I eval the code and press down, it just
inserts alpha. Even though minibuffer-allow-text-properties is set to t
when I look it up. How can that be?

**Comment by minad** [comment]
Alexander Prähauser ***@***.***> writes:

> Huh, but it doesn't. When I eval the code and press down, it just
> inserts alpha. Even though minibuffer-allow-text-properties is set to t
> when I look it up. How can that be?

It inserts alpha, but colored in red, this means with text properties.

**Comment by nameiwillforget** [comment]
"Daniel Mendler" ***@***.***> writes:

> Alexander Prähauser ***@***.***> writes: 
>
>> Huh, but it doesn't. When I eval the code and press down, it just 
>> inserts alpha. Even though minibuffer-allow-text-properties is set to t 
>> when I look it up. How can that be? 
>
> It inserts alpha, but colored in red, this means with text properties. 
>
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you authored the thread.

Oh, I see, sorry. Anyway, I'll find out what exactly causes turning them
on and write the maintainer of that package. Thanks!
